src/kakarot/instructions/memory_operations.cairo

```diff
@@ -434,30 +375,15 @@ namespace MemoryOperations {
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
     }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        // See https://docs.cairo-lang.org/0.12.0/how_cairo_works/functions.html#retrieving-registers
         alloc_locals;
-        let stack = ctx.stack;
-
-        // ------- 1. Get starknet address
-        let starknet_contract_address: felt = ctx.starknet_contract_address;
-
-        // ----- 2. Pop 1 value: key
-
-        // Stack input:
-        // key: key of memory.
-        let (stack, local key) = Stack.pop(stack);
-        // local value: Uint256;
-        // 3. Get the data from storage
-
-        let (local value: Uint256) = IContractAccount.storage(
-            contract_address=starknet_contract_address, key=key
-        );
-
-        let stack: model.Stack* = Stack.push(stack, value);
-
-        // Update context stack.
+        let fp_and_pc = get_fp_and_pc();
+        local __fp__: felt* = fp_and_pc.fp_val;
```

Comment:
where do we use fp here?

Reply:
so this is a bit trickier but it's used in hints that are used under the hood 😱
you can read more at https://docs.cairo-lang.org/0.12.0/how_cairo_works/functions.html#retrieving-registers

practically speaking, when you want to get references of object from the same memory segment, you need this. A MWE is

```
%builtins range_check

from starkware.cairo.common.uint256 import Uint256, uint256_add
from starkware.cairo.common.registers import get_fp_and_pc


struct Transfer {
    sender: felt,
    recipient: felt,
    amount: Uint256*,
}

func foo() -> (res: Uint256) {
    let res = Uint256(1, 2);
    return (res=res);
}

func main{range_check_ptr}() {
    alloc_locals;
    // See https://docs.cairo-lang.org/0.12.0/how_cairo_works/functions.html#retrieving-registers
    let fp_and_pc = get_fp_and_pc();
    local __fp__: felt* = fp_and_pc.fp_val;
    let (local amount) = foo();
    let amount_ptr = &amount;
    tempvar transfer = new Transfer(0x12, 0x34, amount_ptr);
    return ();
}
```

but this doesn't

```
func main{range_check_ptr}() {
    alloc_locals;
    let (local amount) = foo();
    let amount_ptr = &amount;
    tempvar transfer = new Transfer(0x12, 0x34, amount_ptr);
    return ();
}
```

> Using the value of fp directly, requires defining a variable named __fp__.